conduit/data/datasets/vision/nico_plus_plus.py

```diff
@@ -152,90 +139,102 @@ def __init__(
             .drop_duplicates()
             .itertuples(index=False, name=None)
         )
 
     def _check_unzipped(self) -> bool:
-        return all((self._base_dir / attr).exists() for attr in NicoPPAttr)
-
-    def _extract_metadata(self) -> None:
-        """Extract concept/context/superclass information from the image filepaths and it save to csv."""
-        self.logger.info("Extracting metadata.")
-        image_paths: List[Path] = []
-        image_paths.extend(self._base_dir.glob(f"**/*.jpg"))
-        image_paths_str = [str(image.relative_to(self._base_dir)) for image in image_paths]
-        filepaths = pd.Series(image_paths_str)
-        metadata = filepaths.str.split("/", expand=True).rename(
-            columns={0: "subclass", 1: "superclass", 2: "filename"}
+        if not all(
+            (self._base_dir / "public_dg_0416" / "train" / attr).exists() for attr in NicoPPAttr
+        ):
+            return False
+        if not (self._base_dir / "dg_label_id_mapping.json").exists():
+            return False
+        return True
+
+    def _extract_metadata(self):
+        self.logger.info("Generating metadata for NICO++...")
+        attributes = ["autumn", "dim", "grass", "outdoor", "rock", "water"]  # 6 attrs, 60 labels
+        meta = json.load(open(self._base_dir / "dg_label_id_mapping.json", "r"))
+
+        def _make_balanced_testset(
+            df: pd.DataFrame, seed: int, num_samples_val_test: int, verbose: bool = True
+        ) -> pd.DataFrame:
+            # each group has a test set size of (2/3 * num_samples_val_test) and a val set size of
+            # (1/3 * num_samples_val_test); if total samples in original group < num_samples_val_test,
+            # val/test will still be split by 1:2, but no training samples remained
+            import random
+
+            random.seed(seed)
+            val_set, test_set = [], []
+            for g in pd.unique(df["g"]):
+                df_group = df[df["g"] == g]
+                curr_data = df_group["filename"].values
+                random.shuffle(curr_data)
+                split_size = min(len(curr_data), num_samples_val_test)
+                val_set += list(curr_data[: split_size // 3])
+                test_set += list(curr_data[split_size // 3 : split_size])
+            if verbose:
+                print(f"Val: {len(val_set)}\nTest: {len(test_set)}")
+            assert len(set(val_set).intersection(set(test_set))) == 0
+            combined_set = dict(zip(val_set, [1 for _ in range(len(val_set))]))
+            combined_set.update(dict(zip(test_set, [2 for _ in range(len(test_set))])))
+            df["split"] = df["filename"].map(combined_set)
+            df["split"].fillna(0, inplace=True)
+            df["split"] = df.split.astype(int)
+            return df
+
+        all_data = []
+        for c, attr in enumerate(attributes):
+            for label in meta:
+                folder_path = self._base_dir / "public_dg_0416" / "train" / attr / label
```

Comment:
Maybe make "public_dg_0416" a class variable?

Reply:
something like

```python
    subpath: ClassVar[Path] = Path("public_dg_0416") / "train"
```
?

Reply:
Sure, if only the full subpath is ever needed (which seems to be the case)